Create rule S6382[terraform]: Disabling certificate-based authentication is security-sensitive (#594)
SonarSource/rspec: rules/S6382/terraform/rule.adoc

Disabling certificate-based authentication can reduce an organization's ability to react against attacks on its critical functions and data, if any.

> Depending on the Azure resource, multiple authentication choices are possible: Anonymous connections, Basic authentication, or password-based authentication.

Reviewer: This paragraph could start directly by listing the various authentication mechanisms Azure offers:
> Azure offer various authentication options to access resources: Anonymous connections, Basic authentication, password-based authentication or certificate-based authentication.

Reviewer 2: I struggle to see how this suggestion differs from the current one, to be honest

Reviewer 2: I'm working on the phrasing to reduce the sentence size

Reviewer 2: your suggestion is concise and clearer, I'm taking it

> Choosing authentication certificates helps bring client/host trust. 

Reviewer: This is a bit short. Maybe developing a bit more how client certificate authentication would help understand how it brings "client/host trust"

* This Azure resource is essential for mission-critical functions.
* Compliance policies require access to this resource to specifically be authenticated with certificates.

There is a risk if you answered yes to any of those questions.

== Recommended Secure Coding Practices

Enable certificate-based authentication in this Azure resource.

== Sensitive Code Example

For https://azure.microsoft.com/en-us/services/app-service/[App Service]:

----
resource "azurerm_app_service" "example" {
  client_cert_enabled = false # Sensitive
----

For https://azure.microsoft.com/en-us/services/functions/[Function Apps]:

----
resource "azurerm_function_app" "example" {
  client_cert_mode = "Optional" # Sensitive
}
----

For https://azure.microsoft.com/en-us/services/data-factory/Data[Factory Linked Services]:

----
resource "azurerm_data_factory_linked_service_web" "example" {
  authentication_type = "Basic" # Sensitive
}
----

For https://azure.microsoft.com/en-us/services/logic-apps/[Logic App Standard definitions] and https://azure.microsoft.com/en-us/services/api-management/[API Management]:

----
resource "azurerm_api_management" "example" {

>   sku_name = "Consumption_1"

Reviewer: ```suggestion
  sku_name = "example"
```

Reviewer: Please ignore this suggestion. The implementation clearly explain how the `sku_name` influence the detection.

  client_certificate_mode = "Required"
}
----

== Compliant Solution

For https://azure.microsoft.com/en-us/services/app-service/[App Service]:

----
resource "azurerm_app_service" "example" {
  client_cert_enabled = true
----

For https://azure.microsoft.com/en-us/services/functions/[Function Apps]:

----
resource "azurerm_function_app" "example" {
  client_cert_mode = "Required"
}
----

For https://azure.microsoft.com/en-us/services/data-factory/Data[Factory Linked Services]:

----
resource "azurerm_data_factory_linked_service_web" "example" {
  authentication_type = "ClientCertificate"
}
----

For https://azure.microsoft.com/en-us/services/logic-apps/[Logic App Standard definitions] and https://azure.microsoft.com/en-us/services/api-management/[API Management]:

----
resource "azurerm_api_management" "example" {

>   sku_name = "Consumption_1"

Reviewer: ```suggestion
  sku_name = "example"
```

Reviewer: Please ignore this suggestion. The implementation clearly explain how the `sku_name` influence the detection.

  client_certificate_mode = "Required"
}
----

== See

> * https://owasp.org/Top10/A01_2021-Broken_Access_Control/[OWASP Top 10 2021 Category A1] - Boken Access Control
> * https://cwe.mitre.org/data/definitions/668.html[MITRE, CWE-668] - Exposure of Resource to Wrong Sphere

Reviewer: ```suggestion
* https://owasp.org/Top10/A01_2021-Broken_Access_Control/[OWASP Top 10 2021 Category A1] - Boken Access Control
* https://owasp.org/www-project-top-ten/2017/A5_2017-Broken_Access_Control[OWASP Top 10 2017 Category A5] - Boken Access Control
* https://cwe.mitre.org/data/definitions/668.html[MITRE, CWE-668] - Exposure of Resource to Wrong Sphere
```

Disabling certificate-based authentication can reduce an organization's ability to react against attacks on its critical functions and data if any.

> Azure offers various authentication options to access resources: Anonymous connections, Basic authentication, password-based authentication and certificate-based authentication.

Reviewer 2: ```suggestion
Azure offers various authentication options to access resources: Anonymous connections, Basic authentication, password-based authentication, and certificate-based authentication.
```

Choosing certificate-based authentication helps bring client/host trust by allowing the host to verify the client and vice versa.
In case of a security incident, certificates help bring investigators traceability and allow security operations teams to react faster (by massively revoking certificates, for example).

== Ask Yourself Whether

* This Azure resource is essential for the information system infrastructure.
* This Azure resource is essential for mission-critical functions.

> * Compliance policies require access to this resource to specifically be authenticated with certificates.

Reviewer 2: ```suggestion
* Compliance policies require access to this resource to be authenticated with certificates specifically.
```

There is a risk if you answered yes to any of those questions.

== Recommended Secure Coding Practices

Enable certificate-based authentication.

== Sensitive Code Example

For https://azure.microsoft.com/en-us/services/app-service/[App Service]:

----
resource "azurerm_app_service" "example" {
  client_cert_enabled = false # Sensitive
}
----

For https://azure.microsoft.com/en-us/services/logic-apps/[Logic App Standards] and https://azure.microsoft.com/en-us/services/functions/[Function Apps]:

----
resource "azurerm_function_app" "example" {

>   client_cert_mode = "Optional" # Sensitive

Reviewer 3: Implementing this rule
and coming to the case of `client_cert_mode = "Optional" # Sensitive`
made me think -- should I be concerned by case sensitivity of the "Optional" value? (i.e. is "Optional" == "OPTIONAL" == "optional" == "oPtIoNaL" ?)

Reviewer 2: I checked, and it appears that this field is case sensitive.

## Mention in terraform code
- https://github.com/hashicorp/terraform-provider-azurerm/blob/01c73506549d640b99c87c4a40daa7dd6656689d/internal/services/appservice/linux_function_app_resource.go#L166-L168
```golang
		"client_certificate_mode": {
			Type:     pluginsdk.TypeString,
			Optional: true,
			Default:  web.ClientCertModeOptional,
			ValidateFunc: validation.StringInSlice([]string{
				string(web.ClientCertModeOptional),
				string(web.ClientCertModeRequired),
				string(web.ClientCertModeOptionalInteractiveUser),
			}, false),
			Description: "The mode of the Function App's client certificates requirement for incoming requests. Possible values are `Required`, `Optional`, and `OptionalInteractiveUser` ",
		},
```

## Mention in Azure SDK
- https://github.com/Azure/azure-sdk-for-go/blob/main/services/web/mgmt/2021-02-01/web/enums.go#L332-L346

```golang
// ClientCertMode enumerates the values for client cert mode.
type ClientCertMode string

const (
	// ClientCertModeOptional ...
	ClientCertModeOptional ClientCertMode = "Optional"
	// ClientCertModeOptionalInteractiveUser ...
	ClientCertModeOptionalInteractiveUser ClientCertMode = "OptionalInteractiveUser"
	// ClientCertModeRequired ...
	ClientCertModeRequired ClientCertMode = "Required"
)

// PossibleClientCertModeValues returns an array of possible values for the ClientCertMode const type.
func PossibleClientCertModeValues() []ClientCertMode {
	return []ClientCertMode{ClientCertModeOptional, ClientCertModeOptionalInteractiveUser, ClientCertModeRequired}
}
```

Reviewer 2: I'm currently looking into the others but they also seem case sensitive

Reviewer 2: Yes, you can continue to work with them as Case Sensitive Strings, in general. The docs seem to have clear enums on every potential value. Thanks for the feedback! I will think about it my following rule specifications 

# Data Factory
- https://github.com/hashicorp/terraform-provider-azurerm/blob/01c73506549d640b99c87c4a40daa7dd6656689d/vendor/github.com/Azure/azure-sdk-for-go/services/datafactory/mgmt/2018-06-01/datafactory/enums.go#L27-L43
```golang
type AuthenticationType string

const (
	// AuthenticationTypeAnonymous ...
	AuthenticationTypeAnonymous AuthenticationType = "Anonymous"
	// AuthenticationTypeBasic ...
	AuthenticationTypeBasic AuthenticationType = "Basic"
	// AuthenticationTypeClientCertificate ...
	AuthenticationTypeClientCertificate AuthenticationType = "ClientCertificate"
	// AuthenticationTypeWebLinkedServiceTypeProperties ...
	AuthenticationTypeWebLinkedServiceTypeProperties AuthenticationType = "WebLinkedServiceTypeProperties"
)

// PossibleAuthenticationTypeValues returns an array of possible values for the AuthenticationType const type.
func PossibleAuthenticationTypeValues() []AuthenticationType {
	return []AuthenticationType{AuthenticationTypeAnonymous, AuthenticationTypeBasic, AuthenticationTypeClientCertificate, AuthenticationTypeWebLinkedServiceTypeProperties}
}
```

}
----

> For https://azure.microsoft.com/en-us/services/data-factory/Data[Data Factory Linked Services]:

Reviewer 2: ```suggestion
For https://azure.microsoft.com/en-us/services/data-factory/[Data Factory Linked Services]:
```

----
resource "azurerm_data_factory_linked_service_web" "example" {
  authentication_type = "Basic" # Sensitive
}
----

For https://azure.microsoft.com/en-us/services/api-management/[API Management]:

----
resource "azurerm_api_management" "example" {
  sku_name = "Consumption_1"
  client_certificate_mode = "Optional" # Sensitive
}
----

For https://azure.microsoft.com/fr-fr/services/app-service/containers/[Linux and Windows Web Apps]:

----
resource "azurerm_linux_web_app" "sensitive_web_app" {
  client_cert_enabled = false # Sensitive
}
resource "azurerm_linux_web_app" "sensitive_web_app" {
  client_cert_enabled = true
  client_cert_mode = "Optional" # Sensitive
}
----

== Compliant Solution

For https://azure.microsoft.com/en-us/services/app-service/[App Service]:

----
resource "azurerm_app_service" "example" {
  client_cert_enabled = true
}
----

For https://azure.microsoft.com/en-us/services/logic-apps/[Logic App Standards] and https://azure.microsoft.com/en-us/services/functions/[Function Apps]:

----
resource "azurerm_function_app" "example" {
  client_cert_mode = "Required"
}
----

> For https://azure.microsoft.com/en-us/services/data-factory/Data[Data Factory Linked Services]:

Reviewer 2: ```suggestion
For https://azure.microsoft.com/en-us/services/data-factory/[Data Factory Linked Services]:
```

Disabling certificate-based authentication can reduce an organization's ability to react against attacks on its critical functions and data if any.

Azure offers various authentication options to access resources: Anonymous connections, Basic authentication, password-based authentication, and certificate-based authentication.

Choosing certificate-based authentication helps bring client/host trust by allowing the host to verify the client and vice versa.

> In case of a security incident, certificates help bring investigators traceability and allow security operations teams to react faster (by massively revoking certificates, for example).

Reviewer 2: Thanks!

== Ask Yourself Whether

* This Azure resource is essential for the information system infrastructure.
* This Azure resource is essential for mission-critical functions.

> * Compliance policies require access to this resource to be authenticated with certificates specifically.

Reviewer 2: ```suggestion
* Compliance policies require access to this resource to be authenticated with certificates.
```

> * Compliance policies require access to this resource to be authenticated with certificates specifically.

Reviewer 2: absolutely

** If `client_cert_enabled = false`: Make sure that disabling certificate-based authentication is safe here.
* For `data_factory_linked_service_sftp` and `data_factory_linked_service_web`
** If `authentication_type = "Basic"`: Make sure that disabling certificate-based authentication is safe here.
* For `function_app` and `logic_app_standard`: 
** If `client_cert_mode` is missing: Omitting `client_cert_mode` disables certificate-based authentication. Make sure it is safe here.
** If `client_cert_mode = "Optional"`: Make sure that setting certificate-based authentication as optional is safe here.   
* For `linux_web_app` and `windows_web_app`:
** If `client_cert_enabled` is missing: Omitting `client_cert_enabled` disables certificate-based authentication. Make sure it is safe here.
** If `client_cert_enabled = false`: Make sure that disabling certificate-based authentication is safe here.
** If `client_cert_mode` is missing: Omitting `client_cert_mode` disables certificate-based authentication. Make sure it is safe here.
** If `client_cert_mode = "Optional"`: Make sure that setting certificate-based authentication as optional is safe here.   

> ** If both parameters are missing or non-compliant: Make sure that disabling certificate-based authentication is safe here. 

Reviewer: I have the feeling that specifying only 2 messages was enough here:
* Make sure that disabling certificate-based authentication is safe here.
* Omitting {property_name} disables certificate-based authentication. Make sure it is safe here.

Is there some ambiguity you wanted to overcome by defining the message "case by case"?

Reviewer 2: Yes, for example, we can encounter "client_certificate_enabled" and "client_cert_enabled," and one of them is different

But maybe that I should say something like :

```suggestion
For these resources:

* `api_management`:
* `app_service`:
* `data_factory_linked_service_sftp`
* `data_factory_linked_service_web`
* `linux_web_app` 
* `windows_web_app` (if both parameters are non-compliant, flag `client_cert_enabled` first)

These messages apply:

* If an assignment is missing: Omitting {property_name} disables certificate-based authentication. Make sure it is safe here.
* If the assignment is security-sensitive: Make sure that disabling certificate-based authentication is safe here. 

* For `function_app` and `logic_app_standard`:
** Omitting `client_cert_mode` makes certificate-based authentication optional. Make sure it is safe here.
** Make sure that setting certificate-based authentication as optional is safe here.   
```

wdyt?

Reviewer: Yes, I think it's easier to read and provide the same level of information.
